### PR TITLE
[ML-14300] Add a script that updates `ml-package-versions.yml`

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

Adds a script that update `mlflow/ml-package-versions.yml`.

## How is this patch tested?

(Details)

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/ml-package-versions.yml [hunk 1]
```diff
@@ -212,14 +212,14 @@ fastai-1.x:
...
-    maximum: "1.0.61"
+    maximum: "1.0.61" # IGNORE_UPDATE: 2.x is not supported yet
     requirements: ["scikit-learn"]
     run: |
       pytest tests/fastai/test_fastai_model_export.py --large
 
   autologging:
     minimum: "1.0.60"
-    maximum: "1.0.61"
+    maximum: "1.0.61" # IGNORE_UPDATE: 2.x is not supported yet
```
Comment (WeichenXu123) [inline mlflow/ml-package-versions.yml:222]: Why not we add the IGNORE_UPDATE config by a field like:

autologging:
    minimum: "1.0.60"
    maximum: "1.0.61"
    ignore_update: "2.x"

the comment way looks trigger when parsing.

File: dev/update_ml_package_versions.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        key=re.escape(key), category=category
+    )
+    # Matches the following pattern:
+    #
+    # <key>:
+    #   ...
+    #   <category>:
+    #     ...
+    #     maximum: "1.2.3" # IGNORE_UPDATE
+    return re.search(pattern, src, flags=re.DOTALL).group(1) is not None
```
Comment (WeichenXu123) [inline dev/update_ml_package_versions.py:103]: So the logic is, if detecting "# IGNORE_UPDATE xxxx" exists, then return True ? otherwise return False ?

**Review by WeichenXu123 — APPROVED** [review]

File: mlflow/ml-package-versions.yml (hunk 1 above)
Comment (harupy) [reply mlflow/ml-package-versions.yml:222]: I agree. Adding a new field makes more sense. Let me fix.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@harupy Looks awesome generally! I noticed that this script attempts to update the max tested version beyond the current major version. For example, for `fastai v1`, the max tested version is `1.0.61`, but running this PR attempts to set it to a `2.x` version. Can we fix this before merging?

File: mlflow/ml-package-versions.yml [hunk 3]
```diff
@@ -211,13 +211,16 @@ fastai-1.x:
     pip_release: "fastai"
 
   models:
+    # Avoid updating the maximum version to 2.x which isn't supported yet
+    freeze: True
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:215]: @dbczumar Thanks for [this comment](https://github.com/mlflow/mlflow/pull/4174#pullrequestreview-654748537), I  added this field `freeze`. When this field is present, the script doesn't skips updating the maximum version. WDYT?

Comment (WeichenXu123) [reply mlflow/ml-package-versions.yml]: Nit: call the field "frozen" ?

**Review by WeichenXu123 — APPROVED** [review]

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Comment by harupy** [comment]
@dbczumar Thanks!

**Pull request merged**